File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/InsightsLogic.kt
package com.lextures.android.core.lms

object InsightsLogic {
    const val JOURNAL_MAX_LENGTH = 280

    fun formatHours(hours: Double): String = when {
        hours < 0.1 -> "0"
        hours >= 10 -> "%.0f".format(hours)
        else -> "%.1f".format(hours)
    }

    fun hoursFromSeconds(seconds: Int): Double = seconds / 3600.0

    fun goalProgressPercent(progressHours: Float, goalHours: Float?): Int? {
        if (goalHours == null || goalHours <= 0f) return null
        val pct = ((progressHours / goalHours) * 100f).toInt()
        return pct.coerceIn(0, 100)
    }

    fun maxAllocationMinutes(rows: List<StudyTimeAllocationRow>): Double =
        maxOf(1.0, rows.maxOfOrNull { it.minutes } ?: 1.0)

    fun barWidthPercent(minutes: Double, maxMinutes: Double): Double {
        if (maxMinutes <= 0) return 0.0
        return ((minutes / maxMinutes) * 100.0).coerceIn(0.0, 100.0)
    }

    fun moduleCompletionPercent(snapshot: ModulesProgressSnapshot): Int {
        var total = 0
        var complete = 0
        for (module in snapshot.modules) {
            val items = module.items
            if (!items.isNullOrEmpty()) {
                total += items.size
                complete += items.count { it.complete }
            } else {
                total += 1
                if (module.complete) complete += 1
            }
        }
        if (total == 0) return 0
        return ((complete * 100) / total).coerceIn(0, 100)
    }

    fun journalEntryValid(text: String): Boolean {
        val trimmed = text.trim()
        return trimmed.isNotEmpty() && trimmed.length <= JOURNAL_MAX_LENGTH
    }

    fun latestCoachingTip(response: CoachingTipsResponse): CoachingTip? =
        response.latest ?: response.history.firstOrNull()
}

File: clients/ios/Lextures/Core/LMS/InsightsLogic.swift
import Foundation

enum InsightsLogic {
    static let journalMaxLength = 280

    static func formatHours(_ hours: Double) -> String {
        if hours < 0.1 { return "0" }
        return hours >= 10 ? String(format: "%.0f", hours) : String(format: "%.1f", hours)
    }

    static func hoursFromSeconds(_ seconds: Int) -> Double {
        Double(seconds) / 3600.0
    }

    static func goalProgressPercent(progressHours: Float, goalHours: Float?) -> Int? {
        guard let goalHours, goalHours > 0 else { return nil }
        let pct = Int((Double(progressHours) / Double(goalHours)) * 100.0)
        return min(100, max(0, pct))
    }

    static func maxAllocationMinutes(_ rows: [StudyTimeAllocationRow]) -> Double {
        max(1, rows.map(\.minutes).max() ?? 1)
    }

    static func barWidthPercent(minutes: Double, maxMinutes: Double) -> Double {
        guard maxMinutes > 0 else { return 0 }
        return min(100, max(0, (minutes / maxMinutes) * 100))
    }

    static func moduleCompletionPercent(_ snapshot: ModulesProgressSnapshot) -> Int {
        var total = 0
        var complete = 0
        for module in snapshot.modules {
            if let items = module.items, !items.isEmpty {
                total += items.count
                complete += items.filter(\.complete).count
            } else {
                total += 1
                if module.complete { complete += 1 }
            }
        }
        guard total > 0 else { return 0 }
        return min(100, max(0, (complete * 100) / total))
    }

    static func journalEntryValid(_ text: String) -> Bool {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return !trimmed.isEmpty && trimmed.count <= journalMaxLength
    }

    static func latestCoachingTip(from response: CoachingTipsResponse) -> CoachingTip? {
        response.latest ?? response.history?.first
    }

    static func formatJournalDate(_ iso: String) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = formatter.date(from: iso) ?? ISO8601DateFormatter().date(from: iso)
        guard let date else { return iso }
        return date.formatted(date: .abbreviated, time: .shortened)
    }
}